bgpd/bgp_vty.c

```diff
@@ -660,12 +704,11 @@ static int bgp_clear(struct vty *vty, struct bgp *bgp, afi_t afi, safi_t safi,
 			}
 		}
 
-		if (!peer->afc[afi][safi])
+		ret = bgp_peer_clear(peer, afi, safi, NULL, stype);
+
+		/* if afi/safi not defined for this peer, let caller know */
+		if (ret == 1)
```

Comment:
Can we make bgp_peer_clear to return BGP_SUCCESS for (0) and return BGP_ERR_AF_UNCONFIGURED for this case and remove this extra code ? 

The case where you are returning 1 is actually BGP_ERR_AF_UNCONFIGURED if I understand correctly.

Reply:
I've abstracted out the different clear calls so that it can be called in a loop looking for all afi/safi or all safis for a supplied afi, etc.  I didn't want the error message returned for those that are just looping thru possibilities and wanted to let the user know if they supplied an explicit afi/safi that it's not configured.

Reply:
I understand the abstraction part. The code is specifically returning '1', in the case where it is not looping over. i.e. when looking for a specific afi and safi.  I was talking about only that particular case. In fact, peer_clear_soft already returns these error codes.

Reply:
Here is an example that might help.  If you do "clear bgp ipv4 unicast *", it will loop thru all peers and supply the afi/safi for each peer to the new function..  If I just return the AF_UNCONFIGURED directly, every peer that doesn't have ipv4 unicast activated will return an error message stating that it's not configured.  For the commands issued in a loop, I didn't want the error message returned.  Instead, I returned a 1 and in the one case where I wanted to display the AF_UNCONFIGURED error, I translated the 1 into the appropriate error code.

I'm more than happy to do a different approach that I coded, but that was the motivation to the way I did it.